homework_5/homework5.py

```diff
@@ -0,0 +1,229 @@
+#сбрасываем непересеченные с референсами риды
+
+
+# In[36]:
+
+
+merged[['inter_start','inter_end']] = pd.DataFrame(
+    merged.intersection.tolist(), index=merged.index)
+merged = merged.sort_values(by=['start_x','inter_start','inter_end'])
+#сортируем риды
+
+
+# Команда ниже позволяет смотреть пересекается ли конец контига с началом следующего контига - если пересекается, то они создают непрерывную последовательность. Далее команда считает число пересекающихся друг с другом ридом. 
+
+# In[13]:
+
+
+(merged['inter_start']<=merged['inter_end'].shift(1)).astype(int).sum()
+
+
+# Если все отсортированные контиги принадлежащие одному референсу пересекаются друг с другом и включают в себя начало и конец референса - они полностью покрывает референс. При этом число пересечени в этом случае равно числу контигов-1
+
+# In[59]:
+
+
+overlaps = merged.groupby('start_x',as_index=False)    .apply(lambda x: (x['inter_start']<=x['inter_end'].shift(1)).astype(int).sum())
+overlaps = overlaps.rename(columns = {None:'overlaps'})
+overlaps.head()
+
+
+# Отбираем подходящие контиги: пересечений на 1 меньше, чем контигов, начало и конец референса входят в пересечение
+
+# In[62]:
+
+
+contig_data = merged    .groupby(['start_x','end_x','chromosome','attributes'], as_index=False)    .agg(first_inter_start=("inter_start","min"), 
+         last_inter_end=("inter_end","max"), 
+         all_contigs=('name',lambda x: list(x)), 
+         contigs=('source','count'))
+contig_overlaps = contig_data.merge(overlaps, on='start_x', how='inner')
+start_end = contig_overlaps[(contig_overlaps['start_x']>=contig_overlaps['first_inter_start']) 
+                            & (contig_overlaps['end_x']<=contig_overlaps['last_inter_end'])]
+full_RNAs = start_end[start_end['contigs']-1 == start_end['overlaps']]
+full_RNAs
+
+
+#  # 2. Chart customization
+
+# In[88]:
+
+
+diff_exp = pd.read_csv('diffexpr_data.tsv.gz', sep='\t')
+diff_exp
+
+    
+
+
+# In[126]:
+
+
+def gene_group(logFC, pval_corr):
+    if pval_corr <= 0.05:
+        sign = 'Significantly'
+    else:
+        sign = 'Non-significantly'
+    if logFC >= 0:
+        expr = 'upregulated'
+    else:
+        expr = 'downregulated'
+    return f'{sign} {expr}'
+order = ['Significantly downregulated','Significantly upregulated',
+         'Non-significantly downregulated', 'Non-significantly upregulated']
+
+
+# In[127]:
+
+
+diff_exp['gene_type'] = diff_exp.apply(
+    lambda x: gene_group(x.logFC, x.pval_corr), axis = 1)
+
+
+# In[326]:
+
+
+sign = diff_exp[diff_exp['gene_type'].str.contains('S')]
+sign = sign.sort_values(by='logFC')
+top = pd.concat([sign.head(2), sign.tail(2)])
+top
+
+
+# In[379]:
+
+
+plt.rcParams['mathtext.fontset'] = 'custom'
+plt.rcParams['mathtext.bf'] = 'Arial:italic:bold'
+figure(figsize=(10, 6), dpi=100)
+ax = sns.scatterplot(data = diff_exp, x = 'logFC',
+                     y = 'log_pval', hue = 'gene_type', 
+                     hue_order = order, s = 12, linewidth = 0)
```

Comment:
1. Пробелы!
```suggestion
ax = sns.scatterplot(data=diff_exp, x='logFC',
                     y='log_pval', hue='gene_type', 
                     hue_order=order, s=12, linewidth=0)
```
2. Ниже приведён список не соответствий образцу
    + Логарифм не жирный и не курсивный
    + Стрелки без чёрной обводки
    + Лимиты оси абсцисс не симметричные

Reply:
Но в целом, код графика выглядит довольно компактно и аккуратно, это хорошо)